sympy/solvers/solveset.py

```diff
@@ -125,11 +126,62 @@ def _invert_real(f, g_ys, symbol):
                             symbol)
 
     if isinstance(f, Abs):
-        pos = Interval(0, S.Infinity)
-        neg = Interval(S.NegativeInfinity, 0)
-        return _invert_real(f.args[0],
-                    Union(imageset(Lambda(n, n), g_ys).intersect(pos),
-                          imageset(Lambda(n, -n), g_ys).intersect(neg)), symbol)
+        conditions = S.EmptySet
+        condition_interval = Interval(0, oo)
+
+        if f.is_positive:
+            f_n = f
+        else:
+            f_n = f.args[0]
+        g_x, values = _invert_real(f_n, Union(imageset(Lambda(n, n), g_ys),
+                                   imageset(Lambda(n, -n), g_ys)), symbol)
+
+        if isinstance(values, ConditionSet):
+            values_condition = values.condition
+            if not any(condition_atom.has(symbol) for condition_atom in
+                       values_condition.args[0].args):
+                condition_base_set = g_ys
+                if values_condition.args[1] == condition_interval:
+                    condition_base_set = Union(condition_base_set,
+                                               FiniteSet(*values_condition.args[0]))
+                    conditions = Contains(condition_base_set.args,
+                                          condition_interval,
+                                          evaluate=False)
+                else:
+                    condition = Contains(condition_base_set.args,
+                                         condition_interval,
+                                         evaluate=False)
+                    conditions = And(values_condition, condition)
+            else:
+                pass  # No possible usecase
+            values = values.base_set
+
+        if conditions == S.EmptySet:
```

Comment:
This part here applies only to inverting Abs, so shouldn't it be in the `_solve_abs` routine?